extensions/amp-subscriptions/0.1/deferred-account-flow.js

```diff
@@ -0,0 +1,216 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {Services} from '../../../src/services';
+import {UrlBuilder} from '../../amp-subscriptions/0.1/url-builder';
+import {stringifyForPingback} from './stringify-for-pingback';
+
+const LOCAL_STORAGE_KEYS = {
+  HAS_PUBLISHER_ACCOUNT: 'account-exists-on-publisher-side',
+  HAS_REJECTED_ACCOUNT_CREATION: 'user-rejected-account-creation-request',
+};
+
+/**
+ * AMP-side implementation of deferred account creation, a flow responsible
+ * for reconciling account existance discrepancies across platforms.
+ */
+export class DeferredAccountFlow {
+  /**
+   * @param {!../../../src/service/ampdoc-impl.AmpDoc} ampdoc
+   * @param {string} hasAssociatedAccountUrl
+   * @param {string} accountCreationRedirectUrl
+   * @param {../subscription-platform.SubscriptionPlatform} platform
+   */
+  constructor(
+    ampdoc,
+    hasAssociatedAccountUrl,
+    accountCreationRedirectUrl,
+    platform
+  ) {
+    /** @const @private {!../../../src/service/xhr-impl.Xhr} */
+    this.xhr_ = Services.xhrFor(ampdoc.win);
+    /**
+     * @const @private {function(string):void}
+     * @param {string} url
+     */
+    this.navigateTo_ = (url) => {
+      Services.navigationForDoc(ampdoc).navigateTo(ampdoc.win, url);
+    };
+
+    /** @const @private {!../../../src/service/storage-impl.Storage} */
+    this.storage_ = Services.storageForDoc(ampdoc);
+
+    /** @private @const {?string} */
+    this.hasAssociatedAccountUrl_ = hasAssociatedAccountUrl || null;
+
+    /** @private @const {?string} */
+    this.accountCreationRedirectUrl_ = accountCreationRedirectUrl || null;
+    /**
+     * @private @const {!UrlBuilder}
+     */
+    this.urlBuilder_ = new UrlBuilder(ampdoc, platform.getServiceId());
+
+    /**
+     * @private @const {../subscription-platform.SubscriptionPlatform}
+     */
+    this.platform_ = platform;
+  }
+
+  /**
+   * @return {bool} whether deferred account creation is enabled.
+   */
+  isDeferredAccountCreationEnabled() {
+    return this.hasAssociatedAccountUrl_ && this.accountCreationRedirectUrl_;
+  }
+
+  /**
+   * Get key for the storage with the entitlement as part of the key.
+   *
+   * @param {string} storageKey
+   * @param {!./entitlement.Entitlement} entitlements
+   * @return {!Promise}
+   * @private
+   */
+  getStorageKey_(storageKey, entitlements) {
+    const arrayBuffer = new ArrayBuffer(
+      btoa(JSON.stringify(entitlements.json()))
```

Comment:
Aren't there any time-sensitive values in the `Entitlement` object? Maybe we should just use one field from it that's stable. E.g. `subscriptionToken`?

Reply:
The reason John suggested this is because it will cause us to automatically re-ask for confirmation when the subscription is renewed.

Reply:
Yep. That's why I wonder if just the `subscriptionToken` would suffice. We could even go a step further and only take a limited amount of characters out of it since it should be random enough. But I guess hashing accomplishes the same for us.

Reply:
subscriptionToken is in data, not a top level field of entitlement, right? We did it this way so we wouldn't have to create a custom method for SwG. Unsure whether subscriptionToken is supposed to always be there.

Reply:
The order id changes with each renewal in a swg token.  We may need to do something different if we ever add a non swg version of this.  Hashed because this might be a shared computer and we don't want to leave an actual entitlement in local storage.

A though I just had - Do we need to explicitly check this is a swg entitlement and not a linked account?

Reply:
`subscriptionToken` shouldn't change on normal renewal. It would only change when the the "terms of subscription" changed in any way. Which would probably work nicely for us?

Hashing totally makes sense.

I don't remember that we were checking this. But, yes, we definitely should. The `entitlement.source` should be equal to the the granting service itself.

Reply:
Unsure if this is what you're discussing, but when we get the grant entitlement the code will get the remote platform according to the source, which means the source will by definition be equal to the platform.

The deferredAccount is only implemented for swg, so the others should have a no-op.

Reply:
A service can respond with the entitlement on behalf of another service. E.g. with account linking, SwG can respond on the publisher's behalf. In that case the `source` field will indicate that. PTAL on the actual response types. It will be a simple check.

Reply:
@dvoytenko I'm picking this up now @vderosa  is out on leave.  I think the logic should be 

1. Is DAC configured? if no END
2. Do we have a granting local SUBSCRIBER entitlement? if yes END
3. Do we have a granting entitlement that is from a DAC eligible source (currently only google)? if no END
4. is the subscriptionToken hash already in storage (already run)? if yes END

run dac

store subscriptionToken hash to prevent rerun.

Does that sound about right?

Reply:
Correct. And to clarify "store subscriptionToken" means "store hash of the subscriptionToken", right?